File: app/src/main/java/de/westnordost/streetcomplete/quests/bike_parking_type/AddBikeParkingType.java
package de.westnordost.streetcomplete.quests.bike_parking_type;

import android.os.Bundle;
import android.support.annotation.NonNull;

import java.util.List;
import java.util.Map;

import javax.inject.Inject;

import de.westnordost.streetcomplete.R;
import de.westnordost.streetcomplete.data.osm.SimpleOverpassQuestType;
import de.westnordost.streetcomplete.data.osm.changes.StringMapChangesBuilder;
import de.westnordost.streetcomplete.data.osm.download.OverpassMapDataDao;
import de.westnordost.streetcomplete.quests.AbstractQuestAnswerFragment;

public class AddBikeParkingType extends SimpleOverpassQuestType
{
	@Inject
	public AddBikeParkingType(OverpassMapDataDao overpassServer) { super(overpassServer); }

	@Override protected String getTagFilters()
	{
		return "nodes, ways with amenity=bicycle_parking and access!=private and !bicycle_parking";
	}

	@Override
	public void applyAnswerTo(Bundle answer, StringMapChangesBuilder changes) {
		List<String> values = answer.getStringArrayList(AddBikeParkingTypeForm.OSM_VALUES);
		if(values != null  && values.size() == 1)
		{
			changes.add("bicycle_parking", values.get(0));
		}
	}

	@Override public String getCommitMessage() { return "Add bicycle parking type"; }

	@Override public int getTitle(@NonNull Map<String,String> tags)
	{
		return R.string.quest_bicycle_parking_type_title;
	}

	@Override
	public AbstractQuestAnswerFragment createForm() {
		return new AddBikeParkingTypeForm();
	}

	@Override
	public int getIcon() {
		return R.drawable.ic_quest_bicycle_parking;
	}
}

File: app/src/main/java/de/westnordost/streetcomplete/quests/bike_parking_type/AddBikeParkingType.kt
package de.westnordost.streetcomplete.quests.bike_parking_type

import de.westnordost.streetcomplete.R
import de.westnordost.streetcomplete.data.osm.mapdata.Element
import de.westnordost.streetcomplete.data.osm.mapdata.MapDataWithGeometry
import de.westnordost.streetcomplete.data.osm.mapdata.filter
import de.westnordost.streetcomplete.data.osm.osmquests.OsmFilterQuestType
import de.westnordost.streetcomplete.data.osm.osmquests.Tags
import de.westnordost.streetcomplete.data.user.achievements.QuestTypeAchievement.BICYCLIST

class AddBikeParkingType : OsmFilterQuestType<BikeParkingType>() {

    override val elementFilter = """
        nodes, ways with
          amenity = bicycle_parking
          and access !~ private|no
          and !bicycle_parking
    """
    override val changesetComment = "Add bicycle parking type"
    override val wikiLink = "Key:bicycle_parking"
    override val icon = R.drawable.ic_quest_bicycle_parking
    override val isDeleteElementEnabled = true

    override val questTypeAchievements = listOf(BICYCLIST)

    override fun getTitle(tags: Map<String, String>) = R.string.quest_bicycle_parking_type_title

    override fun getHighlightedElements(element: Element, getMapData: () -> MapDataWithGeometry) =
        getMapData().filter("nodes, ways with amenity = bicycle_parking")

    override fun createForm() = AddBikeParkingTypeForm()

    override fun applyAnswerTo(answer: BikeParkingType, tags: Tags, timestampEdited: Long) {
        tags["bicycle_parking"] = answer.osmValue
    }
}
